rfcs/2020-07-21-2744-vicscript.md

```diff
@@ -0,0 +1,153 @@
+
+### Out of Scope
+
+Event expansion and reduction are within the scope, but routing events is not. For example, it would be possible to express with vicscript that an event should be expanded into several, and then at the sink level filter events (potentially using vicscript again) so that they route to different destinations based on their contents. However, vicscript itself will not have any awareness of sinks, only events.
+
+Joining events are within scope, but aggregation of those events is not. For example, it would be possible to aggregate events using a transaction transform (or wasm, etc) in a form where vicscript can be used to express exactly how the fields of those events should be joined. However, vicscript itself will not have the capability to temporarily store events and has no awareness of delivery guarantees or transactions.
+
+## Guide Level Proposal
+
+The `vicscript` transform allows you to mutate events by defining a sequence of mapping statements. Each mapping statement describes a read-only query operation on the right hand side, and on the left hand side a destination for the query result to be mapped within the resulting event:
+
+```toml
+[transforms.mapper]
+  inputs = [ "foo" ]
+  type = "vicscript"
+  mapping = """
+.foo = "hello"
+.bar = .bar + 10
+.baz_is_large = .baz > 10
+"""
+```
+
+> When executing a Vicscript mapping the source event is never mutated.
+
+The most common query type is a dot separated path, describing how to reach a target field within the input event:
+
+```coffee
+.foo = .bar.baz.0.buz
+```
+
+Path sections that include whitespace or other special characters (including `.`) can be quoted:
+
+```coffee
+.foo = .bar."baz.buz".bev
+
+# Use slash escapes within quotes
+.bar = .buz."fub\"\\fob".fab
+```
+
+And supports coalescing fields within a path with a pipe operator `|`, where if the first value is null or missing then the next value is selected, and so on:
+
+```coffee
+.foo = .bar.(baz | buz).bev
+```
+
+Vicscript supports float, int, boolean, string, null, array and object literals:
+
+```coffee
+.first = 7
+.second = false
+.third = "string"
+.fourth = null
+```
+
+Boolean operators and arithmetic galore:
+
+```coffee
+.is_big = .number > 100
+.multiplied = .number * 7
+```
+
+Perform assignments conditionally with an `if` statement:
+
+```coffee
+.id = .id
+.sorted_foo = if .foo.type() == "array" { .foo.sort() }
+```
+
+```text
+In:  {"id":"first","foo":"not an array"}
+Out: {"id":"first","foo":"not an array"}
+
+In:  {"id":"second","foo":["c","a","d","b"]}
+Out: {"id":"second","foo":["c","a","d","b"],"sorted_foo":["a","b","c","d"]}
+```
+
+Remove fields by assigning them the `delete` keyword:
+
+```text
+.foo = delete
+```
```

Comment:
Everything else here is great, but this still feels funky to me. I understand why it's this way, the left-hand side is the "target" and the right-hand side is the "action", but the traditional function syntax seems more intuitive. Going with our `jq` theme, this would just be [`del(.foo)`](https://gist.github.com/olih/f7437fb6962fb3ee9fe95bda8d2c8fa4#dealing-with-json-objects):

```text
del(.foo)
```

Does this introduce significant complexity? I'm also curious to get @lukesteensen's opinion.

Reply:
Yeah, this is another case where I think we can try to avoid keywords and use something less ambiguous like a builtin function.

Reply:
Got options here, bloblang and IDML use `foo = deleted()`, so it's a function rather than a keyword which we could copy. Implementing deletions as a right-hand query allows you to use it in all the same ways as regular mappings, which means they can be conditional:

```coffee
.content = if .type == "bar" {
  .bar.body
} else {
  deleted()
}
```

And also mapped in an iterator in order to remove elements from arrays, etc:

```
.filtered_foos = .foo.map(if .ele.value < 10 { .ele.value + 10 } else { deleted() })
```

However, iterators aren't part of the RFC scope and if we're keeping the spec minimal then I think it's reasonable not to support those cases. It's easy to add a `del(.foo)` left-hand function, but if we want to enable conditional deletions the same as conditional mappings then we'd need to implement `if` as a statement you can put on the left-hand side:

```coffee
if .type == "bar" {
  del(.bar.id)
}
```

At which point we need to decide whether we bother implementing it as a right-hand expression as they'd be parsed and implemented separately. Doing so is useful for expanding the language later, but not so much if we're keeping the language minimal.

Reply:
I see what you're saying, and it makes sense, but from a UX standpoint I find the following much more intuitive:

```coffee
if .type == "bar" {
  del(.bar.id)
}
```

I feel like I might be suggesting something that is fundamentally at odds with the purpose of this language (performance). If that is the case, then I'd prefer to go with your proposal, otherwise, I find the above clearer. Performance is a key requirement of this language and I don't want to sacrifice that.

Reply:
Ok, apologies for the back and forth. I've been trying to build consensus off line on the syntax proposed here. Just confirming was we settled on for this RFC:

```
.foo = "bar"
del(.baz)
```

This would:

1. Add a root `foo` key with the value `"bar"`.
2. Delete the root `baz` key.